src/Open3D/Visualization/Rendering/Filament/FilamentResourceManager.cpp

```diff
@@ -0,0 +1,718 @@
+
+    if (path) {
+        img = io::CreateImageFromFile(path);
+    } else {
+        utility::LogWarning("Empty path for texture loading provided");
+    }
+
+    return CreateTexture(img);
+}
+
+TextureHandle FilamentResourceManager::CreateTexture(
+        const std::shared_ptr<geometry::Image>& img) {
+    TextureHandle handle;
+    if (img->HasData()) {
+        auto texture = LoadTextureFromImage(img);
+
+        handle = RegisterResource<TextureHandle>(engine_, texture, textures_);
+    }
+
+    return handle;
+}
+
+TextureHandle FilamentResourceManager::CreateTexture(
+        const geometry::Image& image) {
+    TextureHandle handle;
+    if (image.HasData()) {
+        auto copy = std::make_shared<geometry::Image>(image);
+
+        auto texture = LoadTextureFromImage(copy);
+
+        handle = RegisterResource<TextureHandle>(engine_, texture, textures_);
+    }
+
+    return handle;
+}
+
+TextureHandle FilamentResourceManager::CreateTextureFilled(
+        const Eigen::Vector3f& color, size_t dimension) {
+    TextureHandle handle;
+    auto texture = LoadFilledTexture(color, dimension);
+    handle = RegisterResource<TextureHandle>(engine_, texture, textures_);
+
+    return handle;
+}
+
+IndirectLightHandle FilamentResourceManager::CreateIndirectLight(
+        const ResourceLoadRequest& request) {
+    IndirectLightHandle handle;
+
+    if (false == request.path.empty()) {
+        std::vector<char> iblData;
+        std::string errorStr;
+
+        if (utility::filesystem::FReadToBuffer(request.path, iblData,
+                                               &errorStr)) {
+            using namespace filament;
+            // will be destroyed later by image::ktx::createTexture
+            auto* iblKtx = new image::KtxBundle(
+                    reinterpret_cast<std::uint8_t*>(iblData.data()),
+                    iblData.size());
+            auto* iblTexture =
+                    image::ktx::createTexture(&engine_, iblKtx, false);
+
+            filament::math::float3 bands[9] = {};
+            if (!iblKtx->getSphericalHarmonics(bands)) {
+                engine_.destroy(iblTexture);
+                request.errorCallback(
+                        request, 2,
+                        "Failed to read spherical harmonics from ktx");
+                return handle;
+            }
+
+            auto indirectLight = IndirectLight::Builder()
+                                         .reflections(iblTexture)
+                                         .irradiance(3, bands)
+                                         .intensity(30000.f)
+                                         .build(engine_);
+
+            if (indirectLight) {
+                handle = RegisterResource<IndirectLightHandle>(
+                        engine_, indirectLight, ibls_);
+
+                auto hTexture = RegisterResource<TextureHandle>(
+                        engine_, iblTexture, textures_);
+                dependencies_[handle].insert(hTexture);
```

Comment:
how/where does iblTexture gets cleaned up in normal conditions? (same for CreateSkybox)

Reply:
Filament's memory-management model is not ideal, so it forces some convolutions on our part. As near as I can tell, RegisterResource() creates a shared_ptr, which deletes its contents with Filament when it destructs. I've added code to clear the ibls_ and skyboxes_ holding tanks to DestroyAll().